src/ray/raylet/node_manager.cc

```diff
@@ -1723,7 +1723,13 @@ bool NodeManager::AssignTask(const Task &task) {
   std::shared_ptr<Worker> worker = worker_pool_.PopWorker(spec);
   if (worker == nullptr) {
     // There are no workers that can execute this task.
-    if (!spec.IsActorTask()) {
+    const bool should_start_new =
+        spec.IsActorCreationTask() && !spec.DynamicWorkerOptions().empty();
+    if (should_start_new) {
+      if (!worker_pool_.PendingRegistrationForTask(spec.GetLanguage(), spec.TaskId())) {
+        worker_pool_.StartWorkerProcess(spec.GetLanguage(), &spec);
+      }
+    } else if (!spec.IsActorTask()) {
```

Comment:
What about starting the new worker process inside `PopWorker` if there's no idle worker?
Then you don't need to check `should_start_new`. worker pool doesn't need to expose `PendingRegistrationForTask` and `StartWorkerProcess`. Code logic would be more clear.